Reward pool behavior before nodes are staked

**Describe the bug**
Tokens sent to the contract before any nodes have stake remain in the balance, but enabling, staking, or disabling nodes with COMMITTEE_ROLE triggers errors.

**To Reproduce**
Steps to reproduceя one of the behaviors:
1. Deploy contract
2. Ensure nodes have no stake.
3. Send tokens (e.g., 3 FAIR) to contract balance.
4. Stake 1 FAIR on the node.
5. Attempt to enable the node.

**Expected behavior**
Nodes can be enabled, staked and disabled according to contract logic without errors. Rewards distribute properly once nodes are staked, healthy, whitelisted, and enabled.

**Actual behavior**
- After enabling a node without stake and then  staking more than 1 gwei triggers `RoundingErrorTooHigh`
- The node staked with 1 FAIR was disabled, and manual attempts to enable it failed with `RoundingErrorTooHigh`.
- Disabling or removing the node triggers `EnumerableMapNonexistentKey`.
- When a node was enabled without stake, it is possible to stake no more than 1 FAIR Gwei.
- Once a node staked with 1 FAIR Gwei is enabled, it cannot be disabled again due to `EnumerableMapNonexistentKey`.

**Screenshots**
Attempt to call disable and enable on Hoodi blockscout, as it writes the reason for the error
<img width="1109" height="597" alt="Image" src="https://github.com/user-attachments/assets/d0947e57-d413-492e-8c20-2a24007a7192" />

<img width="956" height="588" alt="Image" src="https://github.com/user-attachments/assets/e6ab9e46-bb12-4248-aa28-1adfe2104b76" />

**Additional context**
Staking contract where the issue occurs: http://3.71.27.101:10001/address/0x9EcD93c2cF9E551B3B02939c72D3F515A8cb76B0?tab=txs
Also, staking contract on Hoodi where the same behavior was repeated:
 https://eth-hoodi.blockscout.com/address/0x81F664114eBCC62557255cD6A9E23Fd011386022  

Attempt to call disable via script:
```
self = TxRes hash: None, tx_call_result TxCallResult(status=<TxStatus.FAILED: 0>, error='revert', message='0x02b5668600000000...0000000006', data={'data': '0x02b566860000000000000000000000000000000000000000000000000000000000000006'}), receipt None

    def raise_for_status(self) -> None:
        if self.receipt is not None:
            if self.receipt['status'] == TxStatus.FAILED:
                raise TransactionFailedError(
                    'Tx status is failed', {key: str(value) for key, value in self.receipt.items()}
                )
        elif self.tx_call_result is not None and self.tx_call_result.status == TxStatus.FAILED:
            if self.tx_call_result.error == 'revert':
>               raise DryRunRevertError(self.tx_call_result.message)
E               skale.transactions.exceptions.DryRunRevertError: ('0x02b566860000000000000000000000000000000000000000000000000000000000000006', None)
```

--- a/contracts/Staking.sol
+++ b/contracts/Staking.sol
@@ -179,9 +179,9 @@ contract Staking is AccessManagedUpgradeable, ReentrancyGuardUpgradeable, IStaki
         restricted
         onlyExistingActiveNode(node)
     {
+        _pullReward(node);
         (bool wasDisabled, Fair value) = _disabledNodesBalances.tryGet(node);
         require(wasDisabled, NodeIsNotDisabled(node));
-        _pullReward(node);
         Fair balance = _getTotalBalance();
         _rootFund.supply(
             balance,

--- a/contracts/utils/Fund.sol
+++ b/contracts/utils/Fund.sol
@@ -133,6 +133,10 @@ library FundLibrary {
         _processBalanceChange(fund, balanceBeforeSupply);
         Fair balanceBefore = getBalance(fund, balanceBeforeSupply, holder);
         Credit credits = _toCreditsRoundedDown(fund, balanceBeforeSupply, amount);
+        Fair delayedReward = ZERO_FAIR;
+        if (fund.totalCredits == ZERO_CREDIT) {
+            delayedReward = balanceBeforeSupply;
+        }
         (bool holderExists, Credit holderCredits) = fund.credits.tryGet(holder);
         // If holder does not exist, it is added with the credits.
         // If it does exist, set() must return false and value is updated.
@@ -142,7 +146,7 @@ library FundLibrary {
         }
         fund.lastBalance = balanceBeforeSupply + amount;
         Fair balanceAfter = getBalance(fund, fund.lastBalance, holder);
-        _checkAllowedError(balanceBefore, balanceAfter, amount);
+        _checkAllowedError(balanceBefore, balanceAfter, amount + delayedReward);
     }
 
     function getBalance(
@@ -271,6 +275,12 @@ library FundLibrary {
     {
         if (balance == ZERO_FAIR) {
             return Credit.wrap(Fair.unwrap(amount) * CREDIT_PRECISION);
+        }
+        if (fund.totalCredits == ZERO_CREDIT) {
+            // Balance is positive but amount of shares is still zero.
+            // Reward was received before somebody joined the fund.
+            // Give away the reward to first holder joined because there is no one else.
+            return Credit.wrap(Fair.unwrap(amount + balance) * CREDIT_PRECISION);
         }
         return Credit.wrap(
             Math.mulDiv(

--- a/contracts/utils/Fund.sol
+++ b/contracts/utils/Fund.sol
@@ -70,7 +70,7 @@ library FundLibrary {
         internal
     {
         _processBalanceChange(fund, balanceBeforeClaim);
-        
+
         Credit credits = _toCreditsRoundedUp(fund, balanceBeforeClaim, amount);
         if (fund.ownerCredits < credits) {
             revert NotEnoughFee(_toFairRoundedDown(fund, balanceBeforeClaim, ZERO_CREDIT, fund.ownerCredits));
@@ -131,7 +131,7 @@ library FundLibrary {
         internal
     {
         _processBalanceChange(fund, balanceBeforeSupply);
-        Fair balanceBefore = getBalance(fund, balanceBeforeSupply, holder);
+        Fair holderBalanceBefore = getBalance(fund, balanceBeforeSupply, holder);
         Credit credits = _toCreditsRoundedDown(fund, balanceBeforeSupply, amount);
         Fair delayedReward = ZERO_FAIR;
         if (fund.totalCredits == ZERO_CREDIT) {
@@ -146,7 +146,7 @@ library FundLibrary {
         }
         fund.lastBalance = balanceBeforeSupply + amount;
         Fair balanceAfter = getBalance(fund, fund.lastBalance, holder);
-        _checkAllowedError(balanceBefore, balanceAfter, amount + delayedReward);
+        _checkAllowedError(holderBalanceBefore, balanceAfter, amount + delayedReward);
     }
 
     function getBalance(
@@ -254,7 +254,7 @@ library FundLibrary {
         view
         returns (Credit fee)
     {
-        if (balance > fund.lastBalance) {
+        if (balance > fund.lastBalance && fund.feeRate > 0) {
             Fair balanceChange = balance - fund.lastBalance;
             Fair feeInFair = Fair.wrap(
                 Fair.unwrap(balanceChange) * fund.feeRate / 1000
@@ -324,14 +324,15 @@ library FundLibrary {
         view
         returns (Fair fair)
     {
-        if (fund.totalCredits == ZERO_CREDIT) {
+        Credit totalCreditsWithUncountedFee = fund.totalCredits + uncountedFee;
+        if (totalCreditsWithUncountedFee == ZERO_CREDIT) {
             return ZERO_FAIR;
         }
         return Fair.wrap(
             Math.mulDiv(
                 Fair.unwrap(balance),
                 Credit.unwrap(amount),
-                Credit.unwrap(fund.totalCredits + uncountedFee),
+                Credit.unwrap(totalCreditsWithUncountedFee),
                 Math.Rounding.Floor
             )
         );

--- a/test/Staking.ts
+++ b/test/Staking.ts
@@ -1038,4 +1038,69 @@ describe("Staking", () => {
         await setBalance(await ethers.resolveAddress(staking), ethers.parseEther("2"));
         await status.connect(node.wallet).alive();
     });
+
+    it("should correctly process delayed rewards when fee rate is 0", async () => {
+        const {nodesData: [node], staking} = await registeredOnlyNodes();
+        const amount = ethers.parseEther("1");
+        await staking.connect(node.wallet).setFeeRate(0);
+
+        (await staking.getStakedAmount())
+            .should.be.equal(0n);
+        (await staking.getNodeTotalStake(node.id))
+            .should.be.equal(0n);
+        (await staking.getEarnedFeeAmount(node.id))
+            .should.be.equal(0n);
+
+        // Pay reward to a node without stake
+        await setBalance(await staking.getRewardWallet(node.id), amount);
+
+        (await staking.getStakedAmount())
+            .should.be.equal(0n);
+        (await staking.getNodeTotalStake(node.id))
+            .should.be.equal(amount);
+        (await staking.getEarnedFeeAmount(node.id))
+            .should.be.equal(0n);
+
+        // Stake to the node with reward waiting
+        await staking.stake(node.id, {value: amount});
+
+        (await staking.getStakedAmount())
+            .should.be.equal(amount + amount); // stake + delayed reward
+        (await staking.getNodeTotalStake(node.id))
+            .should.be.equal(amount + amount);
+        (await staking.getEarnedFeeAmount(node.id))
+            .should.be.equal(0n);
+    });
+
+    it("should correctly process delayed rewards when fee rate is greater than 0", async () => {
+        const {nodesData: [node], staking} = await registeredOnlyNodes();
+        const amount = ethers.parseEther("1");
+        await staking.connect(node.wallet).setFeeRate(500); // 50%
+
+        (await staking.getStakedAmount())
+            .should.be.equal(0n);
+        (await staking.getNodeTotalStake(node.id))
+            .should.be.equal(0n);
+        (await staking.getEarnedFeeAmount(node.id))
+            .should.be.equal(0n);
+
+        // Pay reward to a node without stake
+        await setBalance(await staking.getRewardWallet(node.id), amount);
+
+        (await staking.getStakedAmount())
+            .should.be.equal(0n);
+        (await staking.getNodeTotalStake(node.id))
+            .should.be.equal(amount);
+        (await staking.getEarnedFeeAmount(node.id))
+            .should.be.equal(amount);
+
+        await staking.stake(node.id, {value: amount});
+
+        (await staking.getStakedAmount())
+            .should.be.equal(amount);
+        (await staking.getNodeTotalStake(node.id))
+            .should.be.equal(amount + amount);
+        (await staking.getEarnedFeeAmount(node.id))
+            .should.be.equal(amount);
+    });
 });

--- a/test/Staking.ts
+++ b/test/Staking.ts
@@ -1029,4 +1029,13 @@ describe("Staking", () => {
         expect(stakeBefore).to.be.greaterThan(stakeAfter);
         expect(await staking.getEarnedFeeAmount(node.id)).to.be.eql(0n);
     });
+
+    it("should allow alive() to enable node after receiving rewards", async () => {
+        const {nodesData, staking, status} = await whitelistedNodes();
+        const [node] = nodesData;
+        const rewardWallet = await staking.getRewardWallet(node.id);
+        await setBalance(rewardWallet, ethers.parseEther("3"));
+        await setBalance(await ethers.resolveAddress(staking), ethers.parseEther("2"));
+        await status.connect(node.wallet).alive();
+    });
 });

